W-19718954-CH1-HA-DR-guide (#334)
mulesoft/docs-hosting: cloudhub-2/modules/ROOT/pages/ch2-ha-dr.adoc

== High Availability Versus Disaster Recovery

High availability (HA) is the measure of a system’s ability to remain accessible despite a system component failure. You generally implement HA by building in multiple levels of fault tolerance or load balancing capabilities into a system. In CloudHub 2.0, you can achieve high availability by deploying your application with multiple replicas.

Disaster recovery (DR) refers to the process of restoring a system to an acceptable previous state after a natural or man-made disaster, such as flooding, tornadoes, earthquakes, fires, power failures, server failures, and misconfigurations.

While they both increase overall availability, the difference is that with HA there's generally no loss of service. HA retains the service, and DR retains the data, but with DR, there's usually a slight loss of service while the DR plan executes and the system restores.

These key terms are essential to understanding HA and DR strategies in CloudHub 2.0:

> Recovery Time Objective (RTO):: The maximum amount of downtime a business can tolerate. The RTO represents the time it takes for the system to recover after a business disruption.

Reviewer: Flagging in case you didn't mean to use two colons here and in the other bullets.

Reviewer 2: @tasmoore - this is how we code DLs.

Multiple-Replica Deployment:: If an application uses multiple replicas, CloudHub 2.0 deploys these replicas across two or more availability zones (AZs) by default. In case of an AZ failure, CloudHub 2.0 automatically restarts the application in a different AZ to maintain availability.

== Disaster Recovery in CloudHub 2.0

Disaster recovery in CloudHub 2.0 focuses on the process of restoring systems after significant disruptions.

=== Global Distribution

You can deploy applications to CloudHub 2.0 in various global regions, including North America, South America, the European Union, and Asia-Pacific. 

If you have the Global Deployment entitlement, and are in the US Cloud or EU cloud, you can deploy applications to CloudHub 2.0 in more than one runtime plane region. 

You can host integrations in a runtime plane region that is closer to your services to reduce latency.

For a list of available runtime plane regions based on the control plane region where your organization is provisioned, see: xref:ch2-architecture.adoc#regions-and-dns-records[Runtime Plane Regions and DNS Records]. 

> The runtime plane region is the region where you deploy your CloudHub 2.0 applications and create CloudHub 2.0 Private Spaces.

Reviewer: ```suggestion
The runtime plane region is the region where you deploy your CloudHub 2.0 applications and create CloudHub 2.0 private spaces.
```

=== Regional Infrastructure

You can create your private space in a region of your choice. The us-east-1 region, for instance, has more than five availability zones, each with multiple data centers, redundant power sources, and internet connectivity.

=== Customer Responsibility for Disaster Recovery

> If your organization has cross-region DR requirements for your applications, build your applications accordingly, and consider deploying them across multiple regions. You can use a load balancer, either cloud-based or on-premises, for applications deployed to different regions to switch traffic to your backup region as part of your DR strategy.

Reviewer: ```suggestion
If your organization has cross-region DR requirements for your applications, build your applications accordingly and consider deploying them across multiple regions. You can use a load balancer, either cloud-based or on-premises, for applications deployed to different regions to switch traffic to your backup region as part of your DR strategy.
```

For more information about how to deploy for HA and DR strategies, see xref:mule-runtime::hadr-guide.adoc[High Availability and Disaster Recovery].

== Disaster Recovery Use Cases and Architecture for the Runtime Plane

CloudHub 2.0 supports various multi-region disaster recovery strategies based on application impact and statefulness. These strategies typically involve a bring-your-own (BYO) global load balancer to manage traffic distribution and failover across regions.

To decouple your runtime plane application availability from the control plane region, configure your primary and backup runtime plane regions to be different from the region where your Anypoint Platform control plane is located.

> You can use a load balancer, euther cloud-based or on-premises, for applications deployed to different regions to provide a better disaster recovery strategy.

Reviewer: ```suggestion
You can use a load balancer, either cloud-based or on-premises, for applications deployed to different regions to provide a better disaster recovery strategy.
```

> image::hadr-load-balancer.png[Diagram showing basic disaster recovery configuration]

Reviewer: ```suggestion
image::hadr-load-balancer.png["Diagram showing basic disaster recovery configuration"]
```

* You can configure CPU-based Horizontal Pod Autoscaling (HPA) where applicable if available to your organization. For more information, see xref:ch2-configure-horizontal-autoscaling.adoc[].

==== Traffic Management

* Use a BYO global load balancer to route traffic to only one region at a time.
* Perform health checks to determine DR and switch routing.
* For non-HTTP use cases, such as Schedulers, externalize a DR flag in your application and use a Groovy script to enable and disable scheduler flows directly in the runtime plane. This reduces reliance on the control plane to trigger your backup applications via a cold standby approach.

==== Licensing

* For organizations using UBP, this configuration doesn’t consume messages or throughput. The configuration consumes fewer flows if you configure the application with xref:ch2-configure-horizontal-autoscaling.adoc[CPU-based HPA].
* For organizations not using UBP, this configuration consumes lower vCores if the backup deployment is in a scaled-down state before it scales up in response to an outage in the primary region.

==== Recovery Time Objective

> If the backup application is in a fully scaled state, you experience lower disruption in the event of an outage. If the backup application is in a scaled-down state, the RTO depends on the time it takes to scale the application up either through control-plane-triggered redeployments, or through xref:ch2-configure-horizontal-autoscaling.adoc[CPU-based HPA].

Reviewer: ```suggestion
If the backup application is in a fully scaled state, you experience lower disruption in the event of an outage. If the backup application is in a scaled-down state, the RTO depends on the time it takes to scale the application up either through control-plane-triggered redeployments or through xref:ch2-configure-horizontal-autoscaling.adoc[CPU-based HPA].
```

For integrating systems to resend events or data of in-flight transactions, capture recovery points in logs and monitoring. This builds reliability in your applications.

=== Use Case 3: Cold Standby Configuration

==== When to Use

Low-impact, stateless, or stateful applications that tolerate longer downtime during recovery.

==== Application State

* Set up the infrastructure, but configure your application to a stopped state in the passive region.
* Start the application only if you detect an outage.

==== Traffic Management

* Use a BYO global load balancer to route traffic only to the region that is up.

> * Perform health checks to determine DR, and scripts to start your applications in backup state and switch traffic routing.

Reviewer: ```suggestion
* Perform health checks to determine DR, and use scripts to start your applications in backup state and switch traffic routing.
```

* If the control plane is unavailable, you can't start the backup application in the runtime plane.

==== Licensing

> Because the backup application is in a stopped state, this configuration doesn't consume additional vCores, Mule flows, Mule messages or Data throughput until the backup application starts in the event of a disaster.

Reviewer: ```suggestion
Because the backup application is in a stopped state, this configuration doesn't consume additional vCores, Mule flows, Mule messages, or Data throughput until the backup application starts in the event of a disaster.
```

==== Recovery Time Objective

The RTO depends on the time it takes to start applications via the control plane.

== Runtime Plane High Availability and Disaster Recovery Summary

High-Impact Stateless API:: Use an active-active setup with active replicas in two regions and an external load balancer.

High-Impact Stateful API:: Use an external load balancer to route traffic, and perform health checks to determine DR and switch traffic.

Medium-Impact Stateless and Stateful API::
+
* Use cold/warm standby, with applications deployed to both regions.
* Stop or scale down the applications or replicas in the secondary region.

== CloudHub 2.0 High Availability and Disaster Recovery Improvements

CloudHub 2.0 improvements over CloudHub include:

> * The runtime plane region and the control plane region are loosely coupled due to a service-oriented architecture.

Reviewer: The sentence structure of each bullet should flow from the intro. For example:

"Loose coupling of the runtime plane region and control plane region due to a service-oriented architecture."

* Clustering, which allows you to configure HA while running the process on a single primary replica.

> * Horizontal Pod Autoscaling (HPA) for warm standby optimizes usage by elastic scaling if a failover is required.

Reviewer: I'm not sure how to revise this one to match the sentence structure of the other bullets because I'm not clear on the meaning. Maybe one of these options?

Usage optimization by elastic scaling by using Horizontal Pod Autoscaling (HPA) for warm standby if a failover is required.

Optimization of usage by elastic scaling by using Horizontal Pod Autoscaling (HPA) for warm standby if a failover is required.

> * Active/passive applications: Configure schedulers, listening connectors, such as Anypoint MQ listener, DB listener, FTP listeners, and Salesforce listener, to be in a stopped or inactive state. 

Reviewer: I suggest changing this sentence to something like this to match the structure of the other bullets:

Schedulers and listening connectors, such as Anypoint MQ, DB, FTP, and Salesforce, configured to be in a stopped or inactive state.

+
Externalize DR flags and expose REST endpoints to enable and disable flows via Groovy script as needed.

== Disaster Recovery for Regional Services

Certain MuleSoft services are regional, which impacts their DR strategy.

Anypoint MQ:: This regional service runs in the same region as the application and provides OOTB region failover to a fallback region for standard queues. If an application's RTO tolerates processing in-flight messages only after the primary site is backed up, Anypoint MQ failover is sufficient. Otherwise, configure a backup application in an active/passive setup, and resend in-flight events and data to the failover queue during a traffic switch. During backup with VM Queues and Anypoint MQ, data can be lost for a few minutes, depending on the configuration.

> Object Store v2:: This regional service runs in the same region as the application and doesn't provide any region failover functionality. An application deployed in a different region can't access it. Bring your own external storage to access or backup data from different regions. Object Store v2 doesn't retrieve data for regular backups. Caching in the object store, used for token management between calls, is region-specific, and you can't transfer it to a secondary region.

Reviewer: ```suggestion
Object Store v2:: This regional service runs in the same region as the application and doesn't provide any region failover functionality. An application deployed in a different region can't access it. Bring your own external storage to access or back up data from different regions. Object Store v2 doesn't retrieve data for regular backups. Caching in the object store, used for token management between calls, is region-specific, and you can't transfer it to a secondary region.
```

= CloudHub 2.0 High Availability and Disaster Recovery

CloudHub 2.0 provides high availability (HA) and disaster recovery (DR) capabilities for applications and protection against hardware failures.

CloudHub 2.0 leverages Amazon AWS for its cloud infrastructure, so availability is directly dependent on Amazon services. CloudHub 2.0 deployments and availability are region-based and correspond to Amazon regions. If an Amazon region goes down, the applications within that region are unavailable and not automatically replicated in other regions.

> While the control plane is unavailable, the runtime plane continues to locally buffer log data and other telemetry data until control plane availability is restored.

Reviewer 3: Rather than saying control plane is unavailable - I would say something on the lines of

In the event of a network partition between the control plane and runtime plane, application deployments in the runtime plane continue to run and the runtime plane continues to locally buffer log and telemetry data until the control plane availability is restored.

> CloudHub 2.0 leverages external messaging services, such as Anypoint MQ, to ensure message reliability through persistent queues. These external queues are highly available within a region, but they can‌ be inaccessible for short periods, generally seconds to minutes, during a regional outage leading to some data loss. Communication with the queues resumes after the region becomes available.

Reviewer 3: Should be reviewed and signed off by MQ/OSv2 team

Reviewer 3: Also - maybe hyperlink to https://docs.mulesoft.com/mq/mq-failover for greater details?

> Anypoint Object Store v2 enables CloudHub 2.0 applications to store data and states across various components and applications. Object Store v2 is maintained in the same region as the deployed CloudHub 2.0 application. Data persists and becomes available after the region returns to service. However, Object Store v2 is a regional service. It doesn't provide any region failover functionality. You can't access data in Object Store v2 from an application in a different region. Implement your own external storage solution for cross-region data persistence and access. Object Store v2 doesn't provide periodic backups.

Reviewer 3: Should be reviewed by MQ/OSv2 team.

== High Availability Versus Disaster Recovery

> High availability (HA) is the measure of a system’s ability to remain accessible despite a system component failure. You generally implement HA by building in multiple levels of fault tolerance or load balancing capabilities into a system. In CloudHub 2.0, you can achieve high availability by distributing multiple replicas across three Availability Zones (AZs) without requiring additional configuration.

Reviewer 3: Since distributing across AZs is not an action controlled by the customer I would rephrase on the lines of: 

In CloudHub 2.0, you can achieve high availability by deploying your application with multiple replicas.

Disaster recovery (DR) refers to the process of restoring a system to an acceptable previous state after a natural or man-made disaster, such as flooding, tornadoes, earthquakes, fires, power failures, server failures, and misconfigurations.

While they both increase overall availability, the difference is that with HA there's generally no loss of service. HA retains the service, and DR retains the data, but with DR, there's usually a slight loss of service while the DR plan executes and the system restores.

These key terms are essential to understanding HA and DR strategies in CloudHub 2.0:

Recovery Time Objective (RTO):: The maximum amount of downtime a business can tolerate. The RTO represents the time it takes for the system to recover after a business disruption.

Recovery Point Objective (RPO):: The maximum amount of time acceptable for data loss after a disaster. The RPO influences the frequency of data backups.

== High Availability in CloudHub 2.0

CloudHub 2.0 implements high availability through several built-in mechanisms, which MuleSoft manages automatically, and require no additional configuration.

> Redundant Locations:: CloudHub 2.0 works with servers in one location and copies transactions in real time across two or more availability zones in completely different, separated locations. This ensures service continuity if one location fails.

Reviewer 3: Is this referring to the runtime plane or control plane? not sure about the source of this - will discuss on the internal doc

> Regular Switching:: CloudHub 2.0 regularly switches among locations for maintenance, compliance, and disaster recovery purposes. This keeps all locations up-to-date and able to handle the load if required.

Reviewer 3: Is this referring to the runtime plane or control plane? not sure about the source of this - will discuss on the internal doc

> Ingress and Egress Traffic Management:: CloudHub 2.0 uses a sophisticated ingress and egress traffic management system that routes traffic optimally, even during maintenance or unexpected downtimes.

Reviewer 3: Is this referring to the runtime plane or control plane? not sure about the source of this - will discuss on the internal doc

> VPN Tunnel Updates:: During VPN tunnel updates, the system sets a lower outbound multiexit discriminator (MED) value on the other tunnel. This maintains connectivity.

Reviewer 3: Is this the same for CH1 & CH2?

Multiple-Replica Deployment:: If an application uses multiple replicas, CloudHub 2.0 deploys these replicas across two or more availability zones (AZs) by default. In case of an AZ failure, CloudHub 2.0 automatically restarts the application in a different AZ to maintain availability.

== Disaster Recovery in CloudHub 2.0

Disaster recovery in CloudHub 2.0 focuses on the process of restoring systems after significant disruptions.

=== Global Distribution

> You can deploy applications to CloudHub 2.0 in various global regions, including North America, South America, the European Union, and Asia-Pacific. This global distribution helps reduce latency by allowing you to host integrations closer to services.

Reviewer 3: If you have the Global Deployment Entitlement, and are in the US Cloud or EU cloud, you can deploy applications to CloudHub 2.0 in more than one runtime plane region. 

You can host integrations in a runtime plane region that is closer to your services to reduce latency.

> For a list of regions where you can deploy your CloudHub 2.0 applications, see: xref:ch2-architecture.adoc#regions-and-dns-records[Runtime Plane Regions and DNS Records].

Reviewer 3: For a list of available runtime plane regions based on the control plane region where your organization is provisioned, see: xref:ch2-architecture.adoc#regions-and-dns-records[Runtime Plane Regions and DNS Records]. Runtime plane region is where you deploy your CloudHub 2.0 applications, and create CloudHub 2.0 Private Spaces.

=== Regional Infrastructure

> You can create your private space in a region of your choice, and the system provisions the corresponding Elastic Kubernetes Service (EKS) cluster in the private space. The us-east-1 region, for instance, has more than five availability zones, each with multiple data centers, redundant power sources, and internet connectivity.

Reviewer 3: that we use EKS is an internal detail. 
"You can create your private space in a region of your choice" suffices.

=== Disaster Recovery Assessment

> In the event of a disaster impacting an entire AWS region, CloudHub 2.0, in conjunction with AWS, performs an assessment to determine which region has the required capacity for the affected workload.

Reviewer 3: Remove - not sure about source, but reads like CH2 provides a fully managed assessment/DR to a region with 'sufficient capacity'

=== Customer Responsibility for Disaster Recovery

> MuleSoft manages platform HA. If your organization has significant DR requirements, build your applications accordingly and consider spreading them across multiple regions.

Reviewer 3: If your organization has cross-region DR requirements for your applications, build your applications accordingly and consider spreading them across multiple regions. You can use a load balancer (cloud or on-premises) for applications deployed to different regions to allow you to switch traffic to your back up region as part of your disaster recovery strategy.

For more information about how to deploy for HA and DR strategies, see xref:mule-runtime::hadr-guide.adoc[High Availability and Disaster Recovery].

== Disaster Recovery Use Cases and Architecture for the Runtime Plane

> CloudHub 2.0 supports various multi-region disaster recovery strategies based on application impact and statefulness. These strategies typically involve a bring-your-own (BYO) global load balancer to manage traffic distribution and failover across regions.

Reviewer 3: To decouple your runtime plane application availability from the control plane region, consider configuring your primary and backup runtime plane regions to be different than the region where your Anypoint Platform control plane is located.

=== Use Case 1: Active-Active Configuration

==== When to Use

High-impact, stateless applications that require continuous availability with minimal downtime.

==== Application State

> * Set up the infrastructure with backup applications running and fully scaled up.

Reviewer 3: Set up the infrastructure with application deployments in multiple regions. Both these application deployments are running simultaneously and fully scaled up.

> * Organizations using a usage-based pricing (UBP) package can enable Horizontal Pod Autoscaling (HPA) on their applications to scale dynamically in response to DR events.

Reviewer 3: Remove

==== Traffic Management

* Use a BYO global load balancer to distribute traffic between both regions.
* Perform health checks to detect disasters.

> * Traffic continues even if the control plane is down.

Reviewer 3: This is not true, if the runtime plane region is the same as the control plane region and the region goes down.

Running Applications continue to run even if the control plane region is unavailable, as long as the applications are deployed in different regions than the control plane.

> ==== Licensing:

Reviewer 2: ```suggestion
==== Licensing
```

> * For organizations using UBP, this configuration distributes messages and throughput between the two regions. It consumes flows based on the number of running replicas per application deployment.

Reviewer 3: I would Hyperlink to messages, throughput and flows definition in our docs - https://docs.mulesoft.com/general/usage-metrics

Also Mule Messages, Mule Flows and Data throughput

* For organizations not using UBP, this configuration consumes additional vCores.

=== Use Case 2: Warm Standby Configuration

==== When to Use

> Medium-impact stateless or stateful APIs that tolerate brief downtime during failover.

Reviewer 3: Medium-impact stateless APIs that tolerate brief downtime during failover or stateful APIs.

==== Application State

> * Set up the infrastructure, but scale down your applications using low vCores in the passive region.

Reviewer 3: Set up the infrastructure in multiple regions. In the passive region, you can either have your backup applications in a fully scaled up state or deploy it in a scaled down with fewer replicas or smaller replica sizes. Scaled down applications will increase reliance on control plane to redeploy and scale up during an outage in the primary runtime plane region.

> * Scale up the applications only if you detect an outage.

Reviewer 3: Remove

> * If your organization uses UBP, configure your applications with HPA where applicable.

Reviewer 3: You could also choose to use CPU based HPA (hyperlink), if available to your organization and applicable.

==== Traffic Management

* Use a BYO global load balancer to route traffic to only one region at a time.
* Perform health checks to determine DR and switch routing.

> * If you configure HPA, the runtime plane continues to run even if the control plane is down.

Reviewer 3: Remove. There are some caveats here w.r.t API manager, connected apps etc.

> 

Reviewer 3: Add another point under Traffic Management:

* For non-HTTP use cases such as Schedulers, you can externalize a DR flag in your application and use a Groovy script to enable and disable scheduler flows directly in the runtime plane. This reduces reliance on the control plane to trigger your backup application via a Cold Standby approach.

==== Licensing

> * For organizations using UBP, this configuration doesn’t consume messages or throughput. The configuration consumes flows while the application is running, and consumes fewer flows if you configure HPA.

Reviewer 3: For organizations using UBP, this configuration doesn’t consume messages or throughput in the backup application, when it is receiving no traffic.

Replace 
_The configuration consumes flows while the application is running, and consumes fewer flows if you configure HPA._
with: 
The configuration consumes fewer flows if you configure the application with CPU based HPA. (Hyperlink to CH2 Horizontal Pod Autoscaling docs page)

> * For organizations not using UBP, this configuration consumes lower vCores.

Reviewer 3: For organizations not using UBP, this configuration consumes lower vCores if the backup deployment is in a scaled down state before it needs to be scaled up in response to an outage in the primary region.

==== Recovery Time Objective

> The RTO depends on the time it takes to scale applications via HPA. HPA is available only to organizations using UBP.

Reviewer 3: If the backup application is in a fully scaled state, customers achieve lower disruption in the event of an outage. If the backup application is in a scaled down state, the RTO depends on the time it takes to scale the application up either through control plane triggered redeployments, or through CPU based Horizontal Pod Autoscaling (hyperlink to the CPU based HPA docs).

> 

Reviewer 3: Build reliability in your applications. For integrating systems to resend event or data of in-flight transactions, capture recovery points in logs and monitoring.

=== Use Case 3: Cold Standby Configuration

==== When to Use

Low-impact, stateless, or stateful applications that tolerate longer downtime during recovery.

==== Application State

* Set up the infrastructure, but configure your application to a stopped state in the passive region.
* Start the application only if you detect an outage.

> ==== Traffic Management:

Reviewer 2: ```suggestion
==== Traffic Management
```

* Use a BYO global load balancer to route traffic only to the region that is up.
* Use health checks to determine DR, and scripts to start your applications in backup state and switch traffic routing.

> * If the control plane is down, the runtime plane DR is impacted.

Reviewer 3: If the control plane is unavailable, you may not be able to start up the backup application deployment in the runtime plane.

> ==== Licensing

Reviewer 3: Since the backup application is in a stopped state, this configuration does not consume additional vCores, Mule Flows, Mule Messages or Data Throughput until the backup application is started up in the event of a disaster.

* For organizations using UBP, This configuration doesn't consume flows, messages, or throughput.
* For organizations not using UBP, this configuration doesn’t consume vCores if the app is stopped.

==== Recovery Time Objective

The RTO depends on the time it takes to start applications via the control plane.

> === Use Case 4: Hot Standby Configuration

Reviewer 3: Let us entirely remove the Hot Standby use case - will suggest absorbing parts of this into Warm Standby

==== When to Use

High-impact, stateful applications that require immediate failover with minimal data loss.

==== Application State

Set up the infrastructure and keep the backup applications running and fully scaled up.

==== Traffic Management

* Use a BYO global load balancer to route traffic only to one region at a time.
* Perform health checks to determine DR and flag configuration.
* Even if the control plane is down, the runtime plane continues to run.
* For non-HTTP use cases, you can externalize a DR flag and use a Groovy script to enable and disable scheduler flows.

==== Licensing

* For organizations using UBP, this configuration doesn't consume messages or throughput. It consumes flows if the application is running.
* For organizations not using UBP, this configuration consumes additional vCores.

==== Recovery Point

For integrating systems to resend event or data of in-flight transactions, capture recovery points in logs and monitoring.

== Runtime Plane High Availability and Disaster Recovery Summary

High-Impact Stateless API:: Use an active-active setup with active replicas in two regions and an external load balancer.

High-Impact Stateful API::
+

> * Use a hot standby setup with applications deployed and started in both regions but processing requests only in one.

Reviewer 3: Replace hot with warm:

Use a warm standby setup with applications deployed and started in both regions but processing requests only in one.

* Use an external load balancer to route traffic, and perform health checks to determine DR and switch traffic.

Medium-Impact Stateless and Stateful API::
+
* Use cold/warm standby, with applications deployed to both regions.
* Stop or scale down the applications or replicas in the secondary region.

== CloudHub 2.0 High Availability and Disaster Recovery Improvements

CloudHub 2.0 improvements over CloudHub include:

> * Region and control plane are decoupled from the runtime plane.

Reviewer 3: Runtime Plane region and Control Plane region are loosely coupled due to a service oriented architecture.

* Clustering allows you to configure HA while running the process on a single primary replica.

> * Horizontal Pod Autoscaling (HPA) for warm standby optimizes usage by elastic scaling if a failover is required, even if the control plane is down.

Reviewer 3: Horizontal Pod Autoscaling (HPA) for warm standby optimizes usage by elastic scaling if a failover is required.

Remove: even if the control plane is down.

* Active/passive applications: Configure schedulers, listening connectors, such as Anypoint MQ listener, DB listener, FTP listeners, and Salesforce listener, to be in a stopped or inactive state. 
+
Externalize DR flags and expose REST endpoints to enable and disable flows via Groovy script as needed.

== Disaster Recovery for Regional Services

> Certain MuleSoft services are regional, which impacts their DR strategy.

Reviewer 3: To be reviewed by MQ/OS team

Object Store v2:: This regional service runs in the same region as the application and doesn't provide any region failover functionality. An application deployed in a different region can't access it. Bring your own external storage to access or backup data from different regions. Object Store v2 doesn't retrieve data for regular backups. Caching in the object store, used for token management between calls, is region-specific, and you can't transfer it to a secondary region.

== Considerations for Disaster Recovery Implementations

Implementing a Disaster Recovery (DR) strategy for applications deployed to CloudHub 2.0 requires careful consideration of several key factors.

> Control Plane Availability:: The CloudHub 2.0 management UI and control plane are only available in the US East region of North America. You can't deploy applications while the US East region is down.

Reviewer 3: This is not accurate -- control plane is available in us-east-1 in US control plane it is also available in EU Cloud (confirm which region), Canada & Japan.
https://docs.mulesoft.com/hosting-home/#control-plane-and-runtime-plane-support-matrix

Caching:: Use caching in the object store to share the tokens among various calls. Object stores are only region-specific, and you can't transfer them to a secondary region.

Rate-Limiting API Policies:: Rate-limiting policies are currently set at the replica level. If you require rate limiting for in-flight transactions during DR, define custom policies using an external multiregion shared object store. Include a DR playbook with steps to adjust rate limits during and after DR.

> Schedulers:: Many applications run on a scheduled basis to read from databases or files. In this case, schedulers continue to run even if the control plane is down.

Reviewer 3: In this case, schedulers continue to run even if the control plane is briefly unavailable.

Listeners and Watermarks:: For listeners like DB Table Row and Salesforce Events, use an external persistent store, like DB or Redis, across replicas and regions for polling and watermarks. This is crucial only if connectors use watermarks directly or indirectly.

> Recommended Primary Runtime Region:: For DR, set us-east-1 runtime plane as the primary region, and us-east-2 as the hot standby region. Even if us-east-1 is down, applications deployed in us-east-2 continue to run. Use an external load balancer to route traffic to the active region and provide transparent failover.

Reviewer 3: Remove

+

> For a warm standby scenario, use us-east-2 as the primary region, and us-east-1 as the secondary region, with an external load balancer for transparent failover.

Reviewer 3: Remove

> IP Allowlist and Firewall Rules:: There's no impact on IP allowlists and firewall rules during a hot standby DR scenario. For a warm standby scenario, update IP allowlists and firewalls as part of the DR playbook.

Reviewer 3: Remove

== Regional Differences and DR for Specific Regions

> Some regions present unique challenges for multiregion DR.

Reviewer 3: Some control plane regions have limitations for multi-region DR

Control Plane to Runtime Plane Ratio:: Unlike the US (1:12) and EU (1:2) regions, Canada, Japan, and India have a 1:1 control plane region to runtime plane region ratio.

> Cross-Region Disaster Recovery Limitations::

Reviewer 2: ```suggestion
== Cross-Region Disaster Recovery Limitations
```

+
* Cross-region runtime plane DR isn't possible in Canada, Japan, and India.

> * Availability Zones are typically 50 miles apart, not the 200 miles required for a true cross-region failover.

Reviewer 3: Remove

> == Disaster Recovery for the Control Plane

Reviewer 3: Disaster Recovery for the Anypoint Platform Control Plane

> The CloudHub 2.0 control plane's high availability and disaster recovery capabilities depend on cross-availability zone (AZ) redundancy within the same AWS region, rather than providing cross-region failover capabilities. The control plane can handle individual data center failures within a region. However, it can't automatically fail over to a different AWS region if the entire region is down.

Reviewer 3: Anypoint control plane (the control plane is not specific to CloudHub 2.0)

Reviewer 3: On second thoughts - remove this section and add the following:

The Anypoint Control Plane encompasses additional Anypoint services apart from CloudHub 2.0. Refer <https://compliance.salesforce.com/en/documents/a005A00000vvLBjQAM> for more details.

> Recovery time is up to 72 hours, with potential data loss of up to 24 hours during regional outages.

Reviewer 3: Remove

= CloudHub 2.0 High Availability and Disaster Recovery

> :page-alias:

Reviewer 2: Is this here for a reason? If so, which old topic should redirect here?

CloudHub 2.0 provides high availability (HA) and disaster recovery (DR) capabilities for applications and protection against hardware failures.

> CloudHub 2.0 leverages Amazon AWS for its cloud infrastructure, so availability is directly dependent on Amazon's services. CloudHub 2.0 deployments and availability are region-based. CloudHub 2.0 regions, in turn, point to the corresponding Amazon regions. If an Amazon region goes down, the applications within that region are unavailable and not automatically replicated in other regions.

Reviewer 2: ```suggestion
CloudHub 2.0 leverages Amazon AWS for its cloud infrastructure, so availability is directly dependent on Amazon services. CloudHub 2.0 deployments and availability are region-based and correspond to Amazon regions. If an Amazon region goes down, the applications within that region are unavailable and not automatically replicated in other regions.
```

While the control plane is unavailable, the runtime plane continues to locally buffer log data and other telemetry data until control plane availability is restored.

CloudHub 2.0 leverages external messaging services, such as Anypoint MQ, to ensure message reliability through persistent queues. These external queues are highly available within a region, but they can‌ be inaccessible for short periods, generally seconds to minutes, during a regional outage leading to some data loss. Communication with the queues resumes after the region becomes available.

> Anypoint Object Store v2 allows CloudHub 2.0 applications to store data and states across various components and applications. Object Store v2 is maintained in the same region as the deployed CloudHub 2.0 application. Data persists and becomes available after the region returns to service. However, Object Store v2 is a regional service. It doesn't provide any region failover functionality. You can't access data in Object Store v2 from an application in a different region. Implement your own external storage solution for cross-region data persistence and access. Object Store v2 doesn't provide periodic backups.

Reviewer 2: ```suggestion
Anypoint Object Store v2 enables CloudHub 2.0 applications to store data and states across various components and applications. Object Store v2 is maintained in the same region as the deployed CloudHub 2.0 application. Data persists and becomes available after the region returns to service. However, Object Store v2 is a regional service. It doesn't provide any region failover functionality. You can't access data in Object Store v2 from an application in a different region. Implement your own external storage solution for cross-region data persistence and access. Object Store v2 doesn't provide periodic backups.
```

> Anypoint Object Store v2 allows CloudHub 2.0 applications to store data and states across various components and applications. Object Store v2 is maintained in the same region as the deployed CloudHub 2.0 application. Data persists and becomes available after the region returns to service. However, Object Store v2 is a regional service. It doesn't provide any region failover functionality. You can't access data in Object Store v2 from an application in a different region. Implement your own external storage solution for cross-region data persistence and access. Object Store v2 doesn't provide periodic backups.

Reviewer 2: ```suggestion
Object Store v2 allows CloudHub 2.0 applications to store data and states across various components and applications. Object Store v2 is maintained in the same region as the deployed CloudHub 2.0 application. Data persists and becomes available after the region returns to service. However, Object Store v2 is a regional service. It doesn't provide any region failover functionality. You can't access data in Object Store v2 from an application in a different region. Implement your own external storage solution for cross-region data persistence and access. Object Store v2 doesn't provide periodic backups.
```

== High Availability Versus Disaster Recovery

High availability (HA) is the measure of a system’s ability to remain accessible despite a system component failure. You generally implement HA by building in multiple levels of fault tolerance or load balancing capabilities into a system. In CloudHub 2.0, you can achieve high availability by distributing multiple replicas across three Availability Zones (AZs) without requiring additional configuration.

Disaster recovery (DR) refers to the process of restoring a system to an acceptable previous state after a natural or man-made disaster, such as flooding, tornadoes, earthquakes, fires, power failures, server failures, and misconfigurations.

While they both increase overall availability, the difference is that with HA there's generally no loss of service. HA retains the service, and DR retains the data, but with DR, there's usually a slight loss of service while the DR plan executes and the system restores.

> === Key Definitions

Reviewer 2: I think you can remove this heading.

These key terms are essential to understanding HA and DR strategies in CloudHub 2.0:

> Recovery Time Objective (RTO):: This is the maximum amount of downtime a business can tolerate. The RTO represents the time it takes for the system to recover after a business disruption.

Reviewer 2: ```suggestion
Recovery Time Objective (RTO):: The maximum amount of downtime a business can tolerate. The RTO represents the time it takes for the system to recover after a business disruption.
```

> Recovery Point Objective (RPO):: This is the maximum amount of time acceptable for data loss after a disaster. The RPO influences the frequency of data backups.

Reviewer 2: ```suggestion
Recovery Point Objective (RPO):: The maximum amount of time acceptable for data loss after a disaster. The RPO influences the frequency of data backups.
```

Reviewer 2: Good use of DL!

== High Availability in CloudHub 2.0

CloudHub 2.0 implements high availability through several built-in mechanisms, which MuleSoft manages automatically, and require no additional configuration.

> === Redundant Locations

Reviewer 2: I think you can use a DL for these, instead of H3s. Each has just a paragraph or two.

> MuleSoft Anypoint works with instances from one place. This mechanism copies transactions in near real time across two or more availability zones in completely different, separate locations. This ensures service continuity if one location fails.

Reviewer 2: ```suggestion
Anypoint Platform works with instances from one location. This mechanism copies transactions in near real time across two or more availability zones in completely different, separate locations. This ensures service continuity if one location fails.
```

Reviewer 2: I'm not sure if this is correct. There isn't a product or platform called "MuleSoft Anypoint". 

I'm not sure what this means: works with instances from one place.

Does it mean, that an app is in one region or AZ?

=== Regular Switching

> MuleSoft regularly switches among locations for maintenance, compliance, and disaster recovery purposes. This keeps all locations up-to-date and able to handle the load if required.

Reviewer 2: Should "MuleSoft" be "CloudHub 2"?

=== Ingress and Egress Traffic Management

CloudHub 2.0 uses a sophisticated ingress and egress traffic management system that routes traffic optimally, even during maintenance or unexpected downtimes.

=== VPN Tunnel Updates

During VPN tunnel updates, the system sets a lower outbound multiexit discriminator (MED) value on the other tunnel. This maintains connectivity.

=== Out-of-the-Box (OOTB) High Availability

> High availability is possible out-of-the-box by configuring multiple replicas across three availability zones (AZs).

Reviewer 2: Earlier (line 32), you say "no additional configuration". 
Also, doesn't OOTB basically mean "works without configuration"?

== Disaster Recovery in CloudHub 2.0

Disaster recovery in CloudHub 2.0 focuses on the process of restoring systems after significant disruptions.

=== Global Distribution

You can deploy applications to CloudHub 2.0 in various global regions, including North America, South America, the European Union, and Asia-Pacific. This global distribution helps reduce latency by allowing you to host integrations closer to services.

For a list of regions where you can deploy your CloudHub 2.0 applications, see: xref:ch2-architecture.adoc#regions-and-dns-records[Runtime Plane Regions and DNS Records].

=== Regional Infrastructure

You can create your private space in a region of your choice, and the system provisions the corresponding Elastic Kubernetes Service (EKS) cluster in the private space. The us-east-1 region, for instance, has more than five availability zones, each with multiple data centers, redundant power sources, and internet connectivity.

=== Disaster Recovery Assessment

> In the event of a disaster impacting an entire AWS region, MuleSoft, in conjunction with AWS, performs an assessment to determine which region has the required capacity for the affected workload.

Reviewer 2: MuleSoft or CloudHub 2?

=== Customer Responsibility for Disaster Recovery

MuleSoft manages platform HA. If your organization has significant DR requirements, build your applications accordingly and consider spreading them across multiple regions.

For more information about how to deploy for HA and DR strategies, see xref:mule-runtime::hadr-guide.adoc[High Availability and Disaster Recovery].

== Disaster Recovery Use Cases and Architecture for the Runtime Plane

CloudHub 2.0 supports various multi-region disaster recovery strategies based on application impact and statefulness. These strategies typically involve a bring-your-own (BYO) global load balancer to manage traffic distribution and failover across regions.

=== Use Case 1: Active-Active Configuration

> *When to Use:* High-impact, stateless applications that require continuous availability with minimal downtime.

Reviewer 2: ```suggestion
When to Use: High-impact, stateless applications that require continuous availability with minimal downtime.
```

Reviewer 2: Don't use bold for non-UI elements.

> *Application State*:

Reviewer 2: ```suggestion
Application State:
```

Reviewer 2: If you want these to be bold, use a DL.

* The RTO depends on the time it takes to start applications via the control plane.

=== Use Case 4: Hot Standby Configuration

*When to Use:* High-impact, stateful applications that require immediate failover with minimal data loss.

*Application State:*

* Set up the infrastructure and keep the backup applications running and fully scaled up.

*Traffic Management:*

* Use a BYO global load balancer to route traffic only to one region at a time.
* Perform health checks to determine DR and flag configuration.
* Even if the control plane is down, the runtime plane continues to run.
* For non-HTTP use cases, you can externalize a DR flag and use a Groovy script to enable and disable scheduler flows.

*Licensing:*

* For organizations using UBP, this configuration doesn't consume messages or throughput. It consumes flows if the application is running.
* For organizations not using UBP, this configuration consumes additional vCores.

*Recovery Point*

* For integrating systems to resend event or data of in-flight transactions, capture recovery points in logs and monitoring.

== Runtime Plane High Availability and Disaster Recovery Summary

[%header%autowidth.spread]

> |===

Reviewer 2: You could simplify this with a DL instead of a table.

|API Type |Configuration
|High-Impact Stateless API | Use an active-active setup with active replicas in two regions and an external load balancer.
|High-Impact Stateful API a|
* Use a hot standby setup with applications deployed and started in both regions but processing requests only in one.
* Use an external load balancer to route traffic, and perform health checks to determine DR and switch traffic.
|Medium-Impact Stateless and Stateful API a|
* Use cold/warm standby, with applications deployed to both regions.
* Stop or scale down the applications or replicas in the secondary region.
|===

== CloudHub 2.0 High Availability and Disaster Recovery Improvements

> CloudHub 2.0 Improvements over CloudHub include:

Reviewer 2: ```suggestion
CloudHub 2.0 improvements over CloudHub include:
```

> * Region and control plane decoupling from the runtime plane.

Reviewer 2: ```suggestion
* Region and control plane are decoupled from the runtime plane.
```

* Clustering allows you to configure HA while running the process on a single primary replica.
* Horizontal Pod Autoscaling (HPA) for warm standby optimizes usage by elastic scaling if a failover is required, even if the control plane is down.

> * Active/Passive applications: Configure schedulers, listening connectors, such as MQ listener, DB listener, FTP listeners, and Salesforce listener, to be in a stopped or inactive state. 

Reviewer 2: ```suggestion
* Active/passive applications: Configure schedulers, listening connectors, such as Anypoint MQ listener, DB listener, FTP listeners, and Salesforce listener, to be in a stopped or inactive state. 
```

+
Externalize DR flags and expose REST endpoints to enable and disable flows via Groovy script as needed.

== Disaster Recovery for Regional Services

> Certain MuleSoft services are regional. This impacts their DR strategy.

Reviewer 2: ```suggestion
Certain MuleSoft services are regional, which impacts their DR strategy.
```

=== Anypoint MQ

> This is a regional service that runs in the same region as the application. It provides out-of-the-box (OOTB) region failover to a fallback region for standard queues. If an application's RTO tolerates processing in-flight messages only after the primary site is backed up, MQ failover is sufficient. Otherwise, configure a backup application in an Active/Passive setup, and resend in-flight events and data to the failover queue during a traffic switch. During backup with VM Queues and Anypoint MQ, data can be lost for a few minutes, depending on the configuration.

Reviewer 2: ```suggestion
This regional service runs in the same region as the application and provides OOTB region failover to a fallback region for standard queues.

If an application's RTO tolerates processing in-flight messages only after the primary site is backed up, Anypoint MQ failover is sufficient. Otherwise, configure a backup application in an active/passive setup, and resend in-flight events and data to the failover queue during a traffic switch. During backup with VM Queues and Anypoint MQ, data can be lost for a few minutes, depending on the configuration.
```

Reviewer 2: (you already defined OOTB)

> === Object Store v2 (OSv2)

Reviewer 2: ```suggestion
=== Object Store v2
```

> This is a regional service running in the same region as the application. OSv2 doesn't provide any region failover functionality. An application deployed in a different region can't access it. Bring your own external storage to access or backup data from different regions. OSv2 doesn't retrieve data for regular backups. Caching in ObjectStore, used for token management between calls, is region-specific, and you can't transfer it to a secondary region.

Reviewer 2: ```suggestion
This regional service runs in the same region as the application and doesn't provide any region failover functionality.

An application deployed in a different region can't access it. Bring your own external storage to access or backup data from different regions. Object Store v2 doesn't retrieve data for regular backups. Caching in the object store, used for token management between calls, is region-specific, and you can't transfer it to a secondary region.
```

== Considerations for Disaster Recovery Implementations

> Implementing a Disaster Recovery (DR) strategy for MuleSoft applications requires careful consideration of several key factors.

Reviewer 2: I think this should be "Mule applications" or "applications deployed to CloudHub 2.0".

> Implementing a Disaster Recovery (DR) strategy for MuleSoft applications requires careful consideration of several key factors.

Reviewer 2: ```suggestion
Implementing a DR strategy for MuleSoft applications requires careful consideration of several key factors.
```

Reviewer 2: Expand the acronym on first mention only.

> === Control Plane Availability

Reviewer 2: Again, these H3s should really be a DL, since they're just a single paragraph.

The CloudHub 2.0 management UI and control plane are only available in the US East region of North America. You can't deploy applications while the US East region is down.

=== Caching

> Use caching in ObjectStore to store the tokens among various calls. Object stores are only region-specific, and you can't transfer them to a secondary region.

Reviewer 2: ```suggestion
Use caching in the object store to share the tokens among various calls. Object stores are only region-specific, and you can't transfer them to a secondary region.
```

Reviewer 2: Should "store" really be "share"?

> === Rate Limiting API Policies

Reviewer 2: ```suggestion
=== Rate-Limiting API Policies
```

> Rate-limiting policies are currently set at the replica level. If you require rate limiting for in-flight transactions during DR, define custom policies using an external multi-region shared object store. Include a DR playbook with steps to adjust rate limits during and after DR.

Reviewer 2: ```suggestion
Rate-limiting policies are currently set at the replica level. If you require rate limiting for in-flight transactions during DR, define custom policies using an external multiregion shared object store. Include a DR playbook with steps to adjust rate limits during and after DR.
```

=== Schedulers

> Many run on a scheduled basis to read from databases or files. In this case, schedulers continue to run even if the control plane is down.

Reviewer 2: Not sure what the first sentence means.

=== Listeners and Watermarks

For listeners like DB Table Row and Salesforce Events, use an external persistent store, like DB or Redis, across replicas and regions for polling and watermarks. This is crucial only if connectors use watermarks directly or indirectly.

=== Recommended Primary Runtime Region

> For DR, set us-east-1 runtime plane as the primary region, and us-east-2 as the hot standby region. Even if us-east-1 is down, applications deployed in us-east-2 continue to run. Use an external load balancer to route traffic to the active region and provide transparent failover. For a warm standby scenario, use us-east-2 as the primary region, and us-east-1 as the secondary region, with an external load balancer for transparent failover.

Reviewer 2: ```suggestion
For DR, set us-east-1 runtime plane as the primary region, and us-east-2 as the hot standby region. Even if us-east-1 is down, applications deployed in us-east-2 continue to run. Use an external load balancer to route traffic to the active region and provide transparent failover.

For a warm standby scenario, use us-east-2 as the primary region, and us-east-1 as the secondary region, with an external load balancer for transparent failover.
```

=== IP Allowlist and Firewall Rules

There's no impact on IP allowlists and firewall rules during a hot standby DR scenario. For a warm standby scenario, update IP allowlists and firewalls as part of the DR playbook.

== Regional Differences and DR for Specific Regions

> Some regions present unique challenges for multi-region DR.

Reviewer 2: ```suggestion
Some regions present unique challenges for multiregion DR.
```

=== Control Plane to Runtime Plane Ratio

* Unlike the US (1:12) and EU (1:2) regions, Canada, Japan, and India have a 1:1 control plane region to runtime plane region ratio.

=== Cross-Region Disaster Recovery Limitations

* Cross-region runtime plane DR isn't possible in Canada, Japan, and India.
* Availability Zones are typically 50 miles apart, not the 200 miles required for a true cross-region failover.

== Disaster Recovery for the Control Plane

The CloudHub 2.0 control plane's high availability and disaster recovery capabilities depend on cross-availability zone (AZ) redundancy within the same AWS region, rather than providing cross-region failover capabilities. The control plane can handle individual data center failures within a region. However, it can't automatically fail over to a different AWS region if the entire region is down.

Recovery time is up to 72 hours, with potential data loss of up to 24 hours during regional outages.

> == See also

Reviewer 2: ```suggestion
== See Also
```
